backend/app/controllers/spree/admin/variants_controller.rb

```diff
@@ -43,7 +43,11 @@ def redirect_on_empty_option_values
       end
 
       def parent
-        @parent ||= Spree::Product.with_discarded.find_by(slug: params[:product_id])
+        begin
+          @parent ||= Spree::Product.with_deleted.find_by!(slug: params[:product_id])
+        rescue ActiveRecord::RecordNotFound => e
+          resource_not_found(e, flash_class: Spree::Product, redirect_url: admin_products_path)
+        end
```

Comment:
You can rescue this at the method level:

```ruby
def parent
  @parent ||= Spree::Product.with_deleted.find_by!(slug: params[:product_id])
  @product = @parent
rescue ActiveRecord::RecordNotFound => e
  resource_not_found(e, flash_class: Spree::Product, redirect_url: admin_products_path)
end

Reply:
Very cool! Did not know this was possible.